### DISCUSSION TITLE
a transitland graphql client for nodejs

### DISCUSSION THREAD

**Discussion by boredland** [transitland/transitland#360] [Show and tell]
Hi there!

I have been using the cool GraphQL endpoint since the first hour at @ioki-mobility and am very happy that it is now generally available! I took the time to set up an automated release process for a NodeJS client using GenQL, in the hope that even more people will use it.

You can find it [here](https://www.npmjs.com/package/@ioki/transitland-gql-client/) and would be happy if you'd give me some feedback in the [repo](https://github.com/ioki-mobility/transitland-gql-client).

Yours,
Jonas

**Comment by irees** [comment]
Hi Jonas, this is great. Is it for request side only, or does it also turn the GraphQL responses into typed objects? We're also considering a minor API change, changing the `after:` parameter in some resolvers from a bare integer to a scalar called `Cursor` that can be either an integer or a string. We'll let you know so you can test your code before deploying this change.

**Reply by boredland** [reply]
> Is it for request side only, or does it also turn the GraphQL responses into typed objects?

It is for both sides. The requests and the responses are fully typed.


https://user-images.githubusercontent.com/4662748/172248996-a4dcdde7-fbf9-449d-bce3-6c84d4c6f54a.mp4

**Reply by boredland** [reply]
> We're also considering a minor API change, changing the after: parameter in some resolvers from a bare integer to a scalar called Cursor that can be either an integer or a string.

the api client will be updated automatically, I hope. clients may break if you make that a breaking change. I'm currently not sure if I use paginated queries throughout our codebase, but I think we do. But: if if it still can be an integer, it shouldn't break too much.

**Reply by boredland** [reply]
by the way it is pretty sad, that gqlgen is [not supporting batched queries](https://github.com/99designs/gqlgen/discussions/1932).

**Reply by irees** [reply]
I'll leave `after:` as-is then, and use the string cursor (which is really just `feed_version id,entity id` base64 encoded) as a new parameter called `cursor:`. I'll try to avoid making too many breaking changes even if it's just changing the schema type and int still works.

There are several things I'd like gqlgen to do but it's "good enough" 😅 A brief history - I wrote the `transitland-lib` library first as a data loading tool and to learn Go, and I enjoyed writing Go enough that I continued to write `transitland-server` in Go as well, even though we had minor difficulties involved in using a statically typed language for something that works on big complex data structures like GraphQL. gqlgen's code generation made it reasonably tolerable. The [dataloaden](https://github.com/vektah/dataloaden) dataloaders were harder... medium term plan is to replace these with the new generics-based data loaders in [graphl-gophers](https://github.com/graph-gophers/dataloader)

**Reply by boredland** [reply]
My GraphQL experience Server-Side is mostly around Apollo, which afaik is kinda the "reference implementation" and always had all the goodies. But I totally understand that there are benefits being in the GO land.
